README
Purpose

Base library for BBC Audio Toolbox

Contains miscellaneous debug, file and position functions

Dependencies

jsoncpp - https://github.com/open-source-parsers/jsoncpp (see Notes on Windows builds, below)

Files

autogen.sh - simple autotools script
bbcat-common - a set of common files for all bbcat-* libraries (will be installed by CMake to /usr/local/share)
CMakeLists.txt - CMake configuration
configure.ac - configure configuration for autotools
COPYING - information on copying this library
debian/ - Debian control and version information
doxygen.am - Doxygen automake file
doxygen.cfg - Doxygen configuration
m4 - folder for autotools
Makefile.am - makefile for automake
README - this file
src/ - source folder containing C/C++ source and header files
test/ - unit test files

bbcat-common/CMakeLibrary.txt			| Common CMake configuration for the libraries

bbcat-common/CMakeProject.txt			| Common CMake configuration for all projects

bbcat-common/Config.cmake.in			| Input files for CMake .cmake configuration file auto-generation
bbcat-common/ConfigVersion.cmake.in		|

bbcat-common/FindPortAudio.cmake		| Various CMake file utilities
bbcat-common/FindThreads.cmake			|
bbcat-common/FindwxWidgets.cmake		|

bbcat-common/findregisterfunctions.sh	| Shell script to auto-generate the register.cpp file

bbcat-common/include/catch/catch.hpp	| Catch unit testing framework

bbcat-common/pkgconfig.in				| Input file for pkg-config auto-generation

src/3DPosition.cpp                      | 3D position, rotation and transformation classes
src/3DPosition.h                        |

src/BackgroundFile.cpp                  | A class derived from EnhancedFile that allows writing to file in a background thread
src/BackgroundFile.h                    |

src/ByteSwap.cpp                        | Byte swapping for endianness control
src/ByteSwap.h                          |

src/CallbackHook.h                      | A simple callback object for sequenced callbacks

src/CMakeLists.txt						| CMake configuration for source files

src/DistanceModel.cpp                   | A model for level and delay calculations based on distance 
src/DistanceModel.h                     |

src/EnhancedFile.cpp                    | A wrapper for FILE * operations which provides some extra functionality 
src/EnhancedFile.h                      |

src/json.cpp                            | Abstraction and support for JSON
src/json.h                              |

src/LoadedVersions.cpp					| A singleton class to hold a list of the loaded versions of libraries and applications
src/LoadedVersions.h					|

src/LockFreeBuffer.h                    | A simple lock-free circular buffer mechanism

src/Makefile.am                         | Makefile for automake

src/misc.cpp                            | Miscelleanous functions and definitions, especially debugging functions
src/misc.h                              |

src/NamedParameter.cpp					| An interface, template and macro that create a named parameter with an optional default value with JSON support
src/NamedParameter.h					|

src/ObjectRegistry.cpp                  | A register of objects that can create themselves (see SelfRegisteringParametricObject.h)
src/ObjectRegistry.h                    |

src/OSCompiler.h                        | Header for determining compiler and OS

src/ParameterSet.cpp                    | A generic key=value handler with operators and comparators
src/ParameterSet.h                      |

src/PerformanceMonitor.cpp              | A multi-point logging runtime performance monitor (with outputs suitable for gnuplot)
src/PerformanceMonitor.h                |

src/RefCount.h							| A simple ref-counting template that allows easy ref-counting object support

src/SelfRegisteringParametricObject.cpp | A base class for objects that can be created from a textual name and parameters (using ParameterSet objects)
src/SelfRegisteringParametricObject.h   |

src/SystemParameters.cpp				| A global registry for system level parameters and paths
src/SystemParameters.h					|

src/Thread.cpp							| Simple thread class that can run a thread via a derived class or callback
src/Thread.h							|

src/ThreadLock.cpp                      | Thread locking classes
src/ThreadLock.h                        |

src/UDPSocket.cpp                       | Simple UDP transmitter/receiver
src/UDPSocket.h                         |

src/UniversalTime.h                     | A simple fraction based timebase with arbitrary numerator and denominator

src/WindowsNet.h						| Windows networking initialisation

src/Windows_uSleep.cpp					| Windows implementation of usleep()
src/Windows_uSleep.h					|

src/register.cpp						| Registration function (see below)

test/CMakeLists.txt						| CMake configuration for tests

test/Makefile.am						| Makefile for automake 

test/jsontests.cpp						| Tests for JSON

test/stringfromtests.cpp				| Tests for StringFrom() functions

test/testbase.cpp						| Test base file

--------------------------------------------------------------------------------
Initialising the Library (IMPORTANT!)

Compilers are clever and try to remove unused code whilst linking applications.
However, sometimes they are too clever and remove some code that *appears* not
to be used but is.  Uses of the SelfRegisteringParametricObject class may
particularly suffer from this as it appears as though nothing is using the
code.

To stop this, each library includes a register.cpp file which explicitly calls a
set of initilisation and registration functions. This file also calls the
registration functions of any libraries it is dependant upon.

For this to work, any application *must* call the registration function of most
dependant library it uses.

For example, for an application using only this library:
#include <bbcat-base/register.h>

using namespace bbcat;

int main(int argc, char *argv[])
{
  // ensure libraries are set up
  bbcat_register_bbcat_base();
}

register.cpp is included in repo but will be updated by the script
bbcat-common/findregisterfunctions.sh if autotools is used.  The CMake process
is also capable of autogenerating this file if the original in the source
directory does not exist.

--------------------------------------------------------------------------------
Building on Windows (Visual Studio)

Download and install CMake (https://cmake.org/download/) *first*
Download and install git (https://git-scm.com/download/win)

Use git-bash to change to the directory where the libraries are to be cloned to

Clone source code, if necessary:
git clone git@github.com:bbcrd/bbcat-base.git

cd bbcat-base
mkdir build
cd build
cmake -G "Visual Studio 14 2015 Win64" .. && cmake --build . --target INSTALL --config Release
(or whatever version of Visual Studio you are using)

Notes on Windows builds

As there is no standardised directories for cmake files, libraries, etc. the build *assumes* that:
1. Library includes, libs and shared files will be stored in c:\local
2. CMake configuration files will be stored in c:\local
3. http://sourceforge.net/projects/pthreads4w/ has been downloaded and extracted to c:\local\pthreads (rename unzip folder to 'pthreads')
4. For JSON support:
4.1. Download the source code, use CMake to generate a solution and build it as x64/Release (install may fail due to privileges)
4.2. Find the .lib file (e.g. in build\src\lib_json\Release) and copy it to c:/local/lib (creating that directory if necessary)
--------------------------------------------------------------------------------

--------------------------------------------------------------------------------
Building on Mac and Linux

There are two build mechanisms supported: autotools and cmake

autotools:
./autogen.sh && configure && make && sudo make install

cmake:
mkdir build ; cd build ; cmake .. && make && sudo make install
--------------------------------------------------------------------------------

--------------------------------------------------------------------------------
Building without cmake or autotools

Using the libraries with other build environments is possible, simply throw the
files into the build environment BUT certain defines must be set to enable
features.

bbcat-base uses the following defines:
ENABLE_JSON=1 		 - enables json support
OLD_JSON_CPP  		 - define if old version of jsoncpp (version 0.6.0, for example) used
_FILE_OFFSET_BITS=64 - enable 64-bit file operations
INSTALL_PREFIX=...   - locations of installation (e.g. /usr/local, c:/local, etc)
USE_PTHREADS         - define if using pthreads rather than std::thread
--------------------------------------------------------------------------------


debian/changelog
bbcat-base (0.1.2.2) quantal; urgency=low

  * Improved CMake build mechanism
  * Added redirection of debug output to OutputDebugString() in Windows
  * Added explicit library registration function (see README)
  * Improved JSON support using jsoncpp library
  * Improved generic conversion to and from JSON (using ToJSON() and FromJSON() functions)
  * Defaults to stl::thread for thread support (*requires* C++ 11 support)
  * Added support for configurable paths by bbcat.conf in home directory, loaded by SystemParameters class
  * Added generic file finding using search paths (configured by the above)
  * Added unit tests using Catch framework

 -- Richard Day <richard.day@bbc.co.uk>  Tue, 26 Apr 2016 17:00:00 +0100

bbcat-base (0.1.2.1) quantal; urgency=low

  * Improved Evaluate() and StringFrom() to avoid need to know what complete format string is
  * Improved LockFreeBuffer<> to allow safe read-ahead and write-ahead
  * Simplified ParameterSet type handling using templates
  * Changed resolve() to be public member function of UDPSocket
  * Added support for recv from and send to in UDPSocket
  * Other minor bugfixes and improvements

  -- Richard Day <richard.day@bbc.co.uk>  Mon, 21 Dec 2015 14:00:00 +0000

bbcat-base (0.1.2.0) quantal; urgency=low

  * Added Windows build support using CMake
  * Added CMake build support for Mac and Linux
  * Added LoadedVersions singleton, a way to keeping track of the loaded versions of compatible libraries
  * REMOVED CreateStrings()/FreeString() mechanism to prevent string lifetime problems
  * REMOVED unused data piping scheme in misc.cpp
  * Added NamedParameter template to provide named, typed parameter handling
  * Added facility for ParameterSet objects to be constructed via an array or vector of values
  * Added RefCount template to provide simple ref-counting for objects
  * Added Thread class to ease creation of threads
  * Modified BackgroundFile to use Thread class
  * Added GetNanosecondTicks() to misc.h to return nanoseconds since machine powerup
  * Expanded Evaluate()/StringFrom() pairs to more types
  * Added CalcTime()/GenarateTime() for conversion back and forth between ns and hh:mm:ss.SSSSS text forms
  * Added addm<> and subz<> to perform safe adding and subtracting for *unsigned* types
  * Added ConvertList<> for converting a vector of pointers to a vector of pointers of another [compatible] type
  * Added version number and git branch embedding using versions.cpp/h
  * Added mechanism to keep object files in final application, even if nothing references them directly
  * Added 64-bit x 32-bit / 32-bit routine in misc.h
  * Added LockFreeBuffer template for handling circular buffers without locks

 -- Richard Day <richard.day@bbc.co.uk>  Mon, 26 Oct 2015 10:30:00 +0000

bbcat-base (0.1.0.0) quantal; urgency=low

  * Initial release.

 -- Richard Day <richard.day@bbc.co.uk>  Tue, 06 Oct 2013 11:27:00 +0100


src/misc.h
#ifndef __MISC__
#define __MISC__

#include <stdint.h>
#include <stdarg.h>

#include <vector>
#include <string>
#include <algorithm>

#include "OSCompiler.h"

#ifdef TARGET_OS_WINDOWS
#define _WINSOCKAPI_    // stops windows.h including winsock.h
#include "windows.h"
#endif

#ifndef NUMBEROF
#define NUMBEROF(x) (sizeof(x) / sizeof(x[0]))
#endif

#ifndef M_PI
#define M_PI        3.14159265358979323846264338327950288   /* pi             */
#endif

#ifndef M_LN2
#define M_LN2       0.693147180559945309417232121458176568  /* loge(2)        */
#endif

#define IFFID(name) (((uint32_t)name[0] << 24) | ((uint32_t)name[1] << 16) | ((uint32_t)name[2] << 8) | (uint32_t)name[3])

#define UNUSED_PARAMETER(name) ((void)(name))

#define USE_BBC_AUDIOTOOLBOX using namespace bbcat;
#define BBC_AUDIOTOOLBOX_START namespace bbcat {
#define BBC_AUDIOTOOLBOX_END }

typedef int16_t       sint16_t;
typedef int32_t       sint32_t;
typedef int64_t       sint64_t;
typedef signed   int  sint_t;
typedef unsigned int  uint_t;
typedef signed   long slong_t;
typedef unsigned long ulong_t;
typedef signed   long long sllong_t;
typedef unsigned long long ullong_t;

#ifdef TARGET_OS_WINDOWS
#include "Windows_uSleep.h"
#endif

#ifdef COMPILER_GCC
#define PACKEDSTRUCT   struct __attribute__ ((packed))
#define PRINTF_FORMAT  __attribute__ ((format (printf,1,2)))
#define PRINTF_FORMAT2 __attribute__ ((format (printf,2,3)))
#define MEMALIGNED(x, decl)  decl __attribute__ ((aligned (x)))
#else
#define PACKEDSTRUCT   struct
#define PRINTF_FORMAT
#define PRINTF_FORMAT2
#define MEMALIGNED(x, decl)  __declspec(align(x)) decl
#endif

BBC_AUDIOTOOLBOX_START

#ifdef __BYTE_ORDER__
// endianness can be determined at compile time
#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
// use #define to allow compiler to optimize
#define MACHINE_IS_BIG_ENDIAN false
#endif

#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
// use #define to allow compiler to optimize
#define MACHINE_IS_BIG_ENDIAN true
#endif
#else
// endianness cannot be determined at compile time, determine at run-time
// defined in ByteSwap.cpp
extern const bool MACHINE_IS_BIG_ENDIAN;
#endif

typedef void (*DEBUGHANDLER)(const char *str, void *context);

/*--------------------------------------------------------------------------------*/
/** Enable logging to file (in a file returned by GetErrorLoggingFile())
 */
/*--------------------------------------------------------------------------------*/
extern void EnableErrorLogging(bool enable = true);

/*--------------------------------------------------------------------------------*/
/** Return filename used to log errors to
 */
/*--------------------------------------------------------------------------------*/
extern const std::string& GetErrorLoggingFile();

/*--------------------------------------------------------------------------------*/
/** These are called by macros and do not need to be called explicitly
 */
/*--------------------------------------------------------------------------------*/
extern void debug_msg(const char *fmt, ...) PRINTF_FORMAT;
extern void debug_err(const char *fmt, ...) PRINTF_FORMAT;

#define BBCERROR debug_err
#define BBCDEBUG debug_msg

/*--------------------------------------------------------------------------------*/
/** Debug output control
 *
 * Before including any include files from bbcat-*, #define BBCDEBUG_LEVEL to a value
 *
 * The value dictates which debug informatin (in that file) is output:
 * BBCDEBUG_LEVEL  Debug output
 * 0            BBCERROR() and BBCDEBUG() ONLY
 * 1            As above plus BBCDEBUG1(())
 * 2            As above plus BBCDEBUG2(())
 * 3            As above plus BBCDEBUG3(())
 * 4            As above plus BBCDEBUG4(())
 * 5            As above plus BBCDEBUG5(())
 *
 * Note: the double-parentheses are required because they allow debug lines that are
 *       NOT going to be outputted to be converted to (void)0 statements (unlike
 *       lower level controls which will evaluate the entrie string and arguments
 *       and then throw it away)
 *
 * If BBCDEBUG_LEVEL is not defined, it is set to 0
 */
/*--------------------------------------------------------------------------------*/
#ifndef BBCDEBUG_LEVEL
#define BBCDEBUG_LEVEL 0
#endif

#if BBCDEBUG_LEVEL >= 1
#define BBCDEBUG1(x) debug_msg x
#else
#define BBCDEBUG1(x) (void)0
#endif

#if BBCDEBUG_LEVEL >= 2
#define BBCDEBUG2(x) debug_msg x
#else
#define BBCDEBUG2(x) (void)0
#endif

#if BBCDEBUG_LEVEL >= 3
#define BBCDEBUG3(x) debug_msg x
#else
#define BBCDEBUG3(x) (void)0
#endif

#if BBCDEBUG_LEVEL >= 4
#define BBCDEBUG4(x) debug_msg x
#else
#define BBCDEBUG4(x) (void)0
#endif

#if BBCDEBUG_LEVEL >= 5
#define BBCDEBUG5(x) debug_msg x
#else
#define BBCDEBUG5(x) (void)0
#endif

#if BBCDEBUG_LEVEL >= 6
#define BBCDEBUG6(x) debug_msg x
#else
#define BBCDEBUG6(x) (void)0
#endif

#if BBCDEBUG_LEVEL >= 7
#define BBCDEBUG7(x) debug_msg x
#else
#define BBCDEBUG7(x) (void)0
#endif

#if BBCDEBUG_LEVEL >= 8
#define BBCDEBUG8(x) debug_msg x
#else
#define BBCDEBUG8(x) (void)0
#endif

#if BBCDEBUG_LEVEL >= 9
#define BBCDEBUG9(x) debug_msg x
#else
#define BBCDEBUG9(x) (void)0
#endif

// fundamental audio processing unit - don't change without careful consideration!
typedef float Sample_t;

typedef PACKEDSTRUCT
{
  uint8_t b[10];
} IEEEEXTENDED;

#ifndef TARGET_OS_WINDOWS
/*--------------------------------------------------------------------------------*/
/** Return machine time on in milliseconds
 */
/*--------------------------------------------------------------------------------*/
extern ulong_t  GetTickCount();
#endif

/*--------------------------------------------------------------------------------*/
/** Multiply 64-bit unsigned integer by 32-bit fraction without overflow
 */
/*--------------------------------------------------------------------------------*/
extern uint64_t muldiv(uint64_t val, uint32_t mul, uint32_t div);

/*--------------------------------------------------------------------------------*/
/** Return machine time on in nanoseconds
 */
/*--------------------------------------------------------------------------------*/
extern uint64_t GetNanosecondTicks();

extern uint32_t IEEEExtendedToINT32u(const IEEEEXTENDED *num);
extern void     INT32uToIEEEExtended(uint32_t val, IEEEEXTENDED *num);

/*--------------------------------------------------------------------------------*/
/** Set debug handler (replacing printf())
 *
 * @param handler debug handler
 * @param context optional parameter to pass to handler
 *
 */
/*--------------------------------------------------------------------------------*/
extern void SetDebugHandler(DEBUGHANDLER handler, void *context = NULL);

/*--------------------------------------------------------------------------------*/
/** Set error handler (replacing printf())
 *
 * @param handler error handler
 * @param context optional parameter to pass to handler
 *
 */
/*--------------------------------------------------------------------------------*/
extern void SetErrorHandler(DEBUGHANDLER handler, void *context = NULL);

/*--------------------------------------------------------------------------------*/
/** Enable use of OutputDebugString in Windows
 */
/*--------------------------------------------------------------------------------*/
extern void EnableWindowsDebug();

/*--------------------------------------------------------------------------------*/
/** Create indentation string
 *
 * @param indent a string representing one level of indentation (e.g. a tab or spaces)
 * @param count indentation level
 *
 * @return a string containing <count> copies of <indent>
 */
/*--------------------------------------------------------------------------------*/
extern std::string CreateIndent(const std::string& indent, uint_t count);

/*--------------------------------------------------------------------------------*/
/** printf for std::string
 *
 * @param str string to be added to
 * @param fmt printf-style format information
 *
 */
/*--------------------------------------------------------------------------------*/
extern void Printf(std::string& str, const char *fmt, ...) PRINTF_FORMAT2;

/*--------------------------------------------------------------------------------*/
/** vprintf for std::string
 *
 * @param str string to be added to
 * @param fmt printf-style format information
 * @param ap ap_list of arguments
 *
 */
/*--------------------------------------------------------------------------------*/
extern void VPrintf(std::string& str, const char *fmt, va_list ap);

/*--------------------------------------------------------------------------------*/
/** Split a string by a delimiter, allowing for quotes to prevent splitting in the wrong place
 *
 * @param str string split
 * @param list list to be populated
 * @param delim delimiter character
 * @param keepempty true to keep empty strings
 * @param maxstrings if non-zero specifies the maximum number of entries in list
 *
 * @return position in string when scanning stopped
 *
 * @note whitespace is IGNORED!
 */
/*--------------------------------------------------------------------------------*/
extern uint_t SplitString(const std::string& str, std::vector<std::string>& list, char delim = ' ', bool keepempty = false, uint_t maxstrings = 0);

/*--------------------------------------------------------------------------------*/
/** Interpolate current towards target at rate coeff, protecting against denormals
 */
/*--------------------------------------------------------------------------------*/
extern void Interpolate(double& current, double target, double coeff, double limit = 1.0e-7);

/*--------------------------------------------------------------------------------*/
/** 'cout' like debug stream handling
 *
 * Basically use local instance of StringStream instead of 'cout' and then wrap entire line in BBCDEBUG() macro
 *
 * For example:
 * BBCDEBUG2((StringStream() << "Test " << i << " of " << n << ":" << StringStream::eol));
 *
 */
/*--------------------------------------------------------------------------------*/
class StringStream
{
public:
  StringStream();
  ~StringStream();

  StringStream& operator << (const std::string& str);
  StringStream& operator << (const char *str);
  StringStream& operator << (sint_t n);
  StringStream& operator << (uint_t n);
  StringStream& operator << (slong_t n);
  StringStream& operator << (ulong_t n);
  StringStream& operator << (sllong_t n);
  StringStream& operator << (ullong_t n);

  const char *get() const;
  operator const char *() const {return get();}
  void clear();

  static const char *eol;

protected:
  std::string data;
};

extern void debug_msg(StringStream& str);
extern void debug_err(StringStream& str);

template <typename Map>
bool map_compare (Map const &lhs, Map const &rhs)
{
  // No predicate needed because there is operator== for pairs already.
  return ((lhs.size() == rhs.size()) &&
          std::equal(lhs.begin(), lhs.end(), rhs.begin()));
}

/*--------------------------------------------------------------------------------*/
/** Prevent value becoming denormalized (bad for performance!)
 */
/*--------------------------------------------------------------------------------*/
extern float  fix_denormal(float val);
extern double fix_denormal(double val);

/*--------------------------------------------------------------------------------*/
/** Factorial of an unsigned integer.
 */
/*--------------------------------------------------------------------------------*/
extern uint_t factorial(uint_t n);

typedef struct {
  const char *name;
  const char *desc;
} PARAMETERDESC;

/*--------------------------------------------------------------------------------*/
/** Add a list of parameters/controls to a list
 */
/*--------------------------------------------------------------------------------*/
extern void AddParametersToList(const PARAMETERDESC *parameters, uint_t n, std::vector<const PARAMETERDESC *>& list);

/*--------------------------------------------------------------------------------*/
/** Convert dB to gain assuming <=-120 is cut
 */
/*--------------------------------------------------------------------------------*/
extern double dBToGain(double db);

/*--------------------------------------------------------------------------------*/
/** Convert gain to dB to gain (-120 is the minimum, including cut)
 */
/*--------------------------------------------------------------------------------*/
extern double GainTodB(double gain);

/*--------------------------------------------------------------------------------*/
/** Attempt to evaluate values from strings
 *
 * If the string starts with '$' or hex is true, the value is evaluted as hex
 */
/*--------------------------------------------------------------------------------*/
extern bool Evaluate(const std::string& str, bool& val);
extern bool Evaluate(const std::string& str, sint_t& val, bool hex = false);
extern bool Evaluate(const std::string& str, uint_t& val, bool hex = false);
extern bool Evaluate(const std::string& str, slong_t& val, bool hex = false);
extern bool Evaluate(const std::string& str, ulong_t& val, bool hex = false);
extern bool Evaluate(const std::string& str, sllong_t& val, bool hex = false);
extern bool Evaluate(const std::string& str, ullong_t& val, bool hex = false);
extern bool Evaluate(const std::string& str, float& val);
extern bool Evaluate(const std::string& str, double& val);
extern bool Evaluate(const std::string& str, std::string& val);

/*--------------------------------------------------------------------------------*/
/** Convert various types to strings
 *
 * @param fmt format specifier (WITHOUT 'l's) and optionally without types
 *
 * @note example value fmt values:
 *  "" default display ('d', 'u' or 'f'), no field formatting
 *  "x" hex display
 *  "016x" hex display with field size of 16, '0' padded
 *  "0.32" default display ('f') with field size of 32
 *
 * Essentially:
 *  if no type specifier is specified, a default type based on the supplied variable type is appended
 *  the correct number of 'l's are inserted before the type specifier for the supplied variable type
 */
/*--------------------------------------------------------------------------------*/
extern std::string StringFrom(bool val);
extern std::string StringFrom(sint_t val, const char *fmt = "");
extern std::string StringFrom(uint_t val, const char *fmt = "");
extern std::string StringFrom(slong_t val, const char *fmt = "");
extern std::string StringFrom(ulong_t val, const char *fmt = "");
extern std::string StringFrom(sllong_t val, const char *fmt = "");
extern std::string StringFrom(ullong_t val, const char *fmt = "");
extern std::string StringFrom(float val, const char *fmt = "0.32");
extern std::string StringFrom(double val, const char *fmt = "0.32");
extern std::string StringFrom(const std::string& val);
extern std::string StringFrom(const void *val);

/*--------------------------------------------------------------------------------*/
/** Bog-standard string search and replace that *should* be in std::string!
 */
/*--------------------------------------------------------------------------------*/
extern std::string SearchAndReplace(const std::string& str, const std::string& search, const std::string& replace);

/*--------------------------------------------------------------------------------*/
/** Very simple wildcard matching
 *
 * @param pat pattern containing characters and/or '*' / '?'
 * @param str string to match
 *
 * @return true if string matches pattern
 *
 * @note this routine is highly recursive!
 */
/*--------------------------------------------------------------------------------*/
extern bool matchstring(const char *pat, const char *str);

/*--------------------------------------------------------------------------------*/
/** Convert text time to ns time
 *
 * @param t ns time variable to be modified
 * @param str time in text format 'hh:mm:ss.SSSSS'
 *
 * @return true if conversion was successful
 */
/*--------------------------------------------------------------------------------*/
extern bool CalcTime(uint64_t& t, const std::string& str);

/*--------------------------------------------------------------------------------*/
/** Convert ns time to text time
 *
 * @param t ns time
 *
 * @return str time in text format 'hh:mm:ss.SSSSS'
 */
/*--------------------------------------------------------------------------------*/
extern std::string GenerateTime(uint64_t t);

namespace limited
{
  /*--------------------------------------------------------------------------------*/
  /** Safe (limited) addition for *unsigned* types (DOES NOT work for signed types)
   */
  /*--------------------------------------------------------------------------------*/
  template<typename T>
  T addm(const T& a, const T& b)
  {
    T c = a + b;
    return (c >= a) ? c : (T)-1;
  }

  /*--------------------------------------------------------------------------------*/
  /** Safe (limited) subtraction for *unsigned* types (DOES NOT work for signed types)
   */
  /*--------------------------------------------------------------------------------*/
  template<typename T>
  T subz(const T& a, const T& b)
  {
    return (a >= b) ? a - b : 0;
  }

  /*--------------------------------------------------------------------------------*/
  /** Return value limited by range
   */
  /*--------------------------------------------------------------------------------*/
  template<typename T>
  const T& limit(const T& a, const T& b, const T& c)
  {
    return ((a < b) ? b : ((a > c) ? c : a));
  }

  /*--------------------------------------------------------------------------------*/
  /** Return whether value is within a range (inclusive)
   */
  /*--------------------------------------------------------------------------------*/
  template<typename T>
  bool inrange(const T& a, const T& b, const T& c)
  {
    return ((a >= b) && (a <= c));
  }
}

/*--------------------------------------------------------------------------------*/
/** Convert a vector of pointers to a vector of pointers of another [compatible] type
 */
/*--------------------------------------------------------------------------------*/
template<typename T1,typename T2>
std::vector<T1 *> ConvertList(const std::vector<T2 *>& vec) {
  std::vector<T1 *> res(vec.size());
  uint_t i;
  for (i = 0; i < vec.size(); i++) res[i] = vec[i];
  return res;
}

/*--------------------------------------------------------------------------------*/
/** Convert a vector of pointers to a vector of pointers of another [compatible] type
 */
/*--------------------------------------------------------------------------------*/
template<typename T1,typename T2>
void ConvertList(const std::vector<T2 *>& vec, std::vector<T1 *>& res) {
  uint_t i;
  for (i = 0; i < vec.size(); i++) res.push_back(vec[i]);
}

/*--------------------------------------------------------------------------------*/
/** Convert a vector of pointers to a vector of pointers of another [compatible] type
 */
/*--------------------------------------------------------------------------------*/
template<typename T1,typename T2>
void ConvertList(const std::vector<T2 *>& vec, std::vector<const T1 *>& res) {
  uint_t i;
  for (i = 0; i < vec.size(); i++) res.push_back(vec[i]);
}

/*--------------------------------------------------------------------------------*/
/** Find a file using various sources of possible paths
 *
 * @param filename filename of file to search for
 * @param paths list of paths to search separated by ';'
 *
 * @return path of found file or empty if file not found
 */
/*--------------------------------------------------------------------------------*/
extern std::string FindFile(const std::string& filename, const std::string& pathlist = "");

/*--------------------------------------------------------------------------------*/
/** Find a file using various sources of possible paths
 *
 * @param filename filename of file to search for
 * @param paths list of paths to search
 * @param npaths number of paths in above list
 *
 * @return path of found file or empty if file not found
 *
 * @note each entry can be a list of directories separated by ';'
 */
/*--------------------------------------------------------------------------------*/
extern std::string FindFile(const std::string& filename, const char *paths[], uint_t npaths);

/*--------------------------------------------------------------------------------*/
/** Find a file using various sources of possible paths
 *
 * @param filename filename of file to search for
 * @param paths list of paths to search 
 * @param npaths number of paths in above list
 *
 * @return path of found file or empty if file not found
 *
 * @note each entry can be a list of directories separated by ';'
 */
/*--------------------------------------------------------------------------------*/
extern std::string FindFile(const std::string& filename, const std::string *paths, uint_t npaths);

/*--------------------------------------------------------------------------------*/
/** Find a file using various sources of possible paths
 *
 * @param filename filename of file to search for
 * @param paths list of paths to search 
 *
 * @return path of found file or empty if file not found
 *
 * @note each entry can be a list of directories separated by ';'
 */
/*--------------------------------------------------------------------------------*/
extern std::string FindFile(const std::string& filename, const std::vector<std::string>& paths);

/*--------------------------------------------------------------------------------*/
/** Find path that exists and is a directory
 *
 * @param paths list of paths to search separated by ';'
 *
 * @return first existing path found or empty if no path found not found
 */
/*--------------------------------------------------------------------------------*/
extern std::string FindPath(const std::string& pathlist);

/*--------------------------------------------------------------------------------*/
/** Find path that exists and is a directory
 *
 * @param paths list of paths to search
 * @param npaths number of paths
 *
 * @return first existing path found or empty if no path found not found
 */
/*--------------------------------------------------------------------------------*/
extern std::string FindPath(const char *paths[], uint_t npaths);

/*--------------------------------------------------------------------------------*/
/** Find path that exists and is a directory
 *
 * @param paths list of paths to search
 * @param npaths number of paths
 *
 * @return first existing path found or empty if no path found not found
 */
/*--------------------------------------------------------------------------------*/
extern std::string FindPath(const std::string *paths, uint_t npaths);

/*--------------------------------------------------------------------------------*/
/** Find path that exists and is a directory
 *
 * @param paths list of paths to search
 *
 * @return first existing path found or empty if no path found not found
 */
/*--------------------------------------------------------------------------------*/
extern std::string FindPath(const std::vector<std::string>& paths);

/*--------------------------------------------------------------------------------*/
/** Find a directory to write a file and return full filename
 *
 * @param filename list of path/filename combinations to test
 *
 * @return full filename or empty if no path could be found
 */
/*--------------------------------------------------------------------------------*/
extern std::string FindPathForFile(const std::string& filename);

BBC_AUDIOTOOLBOX_END

#endif
